# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/VideoConduit.cpp
# Commit: 26913ec8db0b
# Full Hash: 26913ec8db0b350b29145e9502136b29efdeccbc
# Author: Byron Campen [:bwc] <docfaraday@gmail.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 5
# Description:
#   Bug 1654112 - Reimplement ssrc/csrc stats cache. r=ng
#   
#   The cache can be a lot simpler than the old one, since libwebrtc handles
#   pruning old entries for us, and since we're now telling libwebrtc when we
#   render frames (so we don't have to make our own stats with guesses of when
# ==============================================================================

diff -r 18d3bc91b71e -r 26913ec8db0b dom/media/webrtc/libwebrtcglue/VideoConduit.cpp
--- a/dom/media/webrtc/libwebrtcglue/VideoConduit.cpp	Thu Feb 25 13:07:39 2021 -0600
+++ b/dom/media/webrtc/libwebrtcglue/VideoConduit.cpp	Mon Mar 15 12:19:57 2021 -0500
@@ -868,37 +868,6 @@
   return mCall->Call()->GetStats();
 }
 
-void WebrtcVideoConduit::GetRtpSources(
-    nsTArray<dom::RTCRtpSourceEntry>& outSources) {
-  MOZ_ASSERT(NS_IsMainThread());
-  auto current = TaskQueueWrapper::MainAsCurrent();
-  std::vector<webrtc::RtpSource> sources = mRecvStream->GetSources();
-  outSources.Clear();
-  for (const auto& source : sources) {
-    MOZ_ASSERT(!source.audio_level(), "Video cannot have an audio level");
-    dom::RTCRtpSourceEntry domEntry;
-    domEntry.mSource = source.source_id();
-    switch (source.source_type()) {
-      case webrtc::RtpSourceType::SSRC:
-        domEntry.mSourceType = dom::RTCRtpSourceEntryType::Synchronization;
-        break;
-      case webrtc::RtpSourceType::CSRC:
-        domEntry.mSourceType = dom::RTCRtpSourceEntryType::Contributing;
-        break;
-      default:
-        MOZ_CRASH("Unexpected RTCRtpSourceEntryType");
-    }
-    // Fix up timestamp to be consistent with JS time. We assume that
-    // source.timestamp_ms() was not terribly long ago, and so clock drift
-    // between thje libwebrtc clock and our JS clock is not that significant.
-    double ago = webrtc::Clock::GetRealTimeClock()->TimeInMilliseconds() -
-                 source.timestamp_ms();
-    domEntry.mTimestamp = GetNow() - ago;
-    domEntry.mRtpTimestamp = source.rtp_timestamp();
-    outSources.AppendElement(std::move(domEntry));
-  }
-}
-
 MediaConduitErrorCode WebrtcVideoConduit::InitMain() {
   MOZ_ASSERT(NS_IsMainThread());
 
@@ -1414,6 +1383,7 @@
 
   mVideoBroadcaster.OnFrame(webrtc::VideoFrame(
       buffer, frame.timestamp(), frame.render_time_ms(), frame.rotation()));
+
   return kMediaConduitNoError;
 }
 
@@ -1565,6 +1535,22 @@
   return StartReceivingLocked();
 }
 
+void WebrtcVideoConduit::OnFrameDelivered() {
+  // Spec says to "queue a task" to update contributing/synchronization source
+  // stats; that's what we're doing here.
+  NS_DispatchToMainThread(
+      media::NewRunnableFrom([this, self = RefPtr<WebrtcVideoConduit>(this)]() {
+        auto current = TaskQueueWrapper::MainAsCurrent();
+        std::vector<webrtc::RtpSource> sources;
+        if (mRecvStream) {
+          sources = mRecvStream->GetSources();
+        }
+        UpdateRtpSources(sources);
+        return NS_OK;
+      }),
+      NS_DISPATCH_NORMAL);
+}
+
 MediaConduitErrorCode WebrtcVideoConduit::StopTransmittingLocked() {
   MOZ_ASSERT(NS_IsMainThread());
   auto current = TaskQueueWrapper::MainAsCurrent();